File: examples/files-in-a-sandbox/verify.mjs
/**
 * Proves this example works: boot a sandbox, write a text file, read it back,
 * list the directory, assert the round-trip, dispose.
 * Token/project from env (CI) or ~/.config/tenki/config.yaml (local `tenki login`).
 * Exits non-zero on any failure.
 */
import { TenkiSandbox } from "@tenkicloud/sandbox";
import { readFileSync } from "node:fs";
import { homedir } from "node:os";

const cfg = (key) => {
	try {
		const c = readFileSync(`${homedir()}/.config/tenki/config.yaml`, "utf8");
		return (c.match(new RegExp(`^${key}:\\s*(.+)$`, "m"))?.[1] ?? "").trim();
	} catch {
		return "";
	}
};

const authToken = process.env.TENKI_AUTH_TOKEN || process.env.TENKI_API_KEY || cfg("auth_token");
const projectId = process.env.TENKI_PROJECT_ID || cfg("current_project_id");
const workspaceId = process.env.TENKI_WORKSPACE_ID || cfg("current_workspace_id");
if (!authToken) {
	console.error("No token. Set TENKI_AUTH_TOKEN, or run `tenki login`.");
	process.exit(1);
}

const path = "notes.txt";
const text = "hello from tenki\n";

const tenki = new TenkiSandbox({ authToken });
let sandbox;
try {
	sandbox = await tenki.createAndWait({ cpuCores: 1, memoryMb: 1024, projectId, workspaceId });

	// write -> read back -> assert the bytes survive the round-trip.
	await sandbox.writeFile(path, text);
	const bytes = await sandbox.readFile(path);
	const back = new TextDecoder().decode(bytes);
	if (back !== text) {
		throw new Error(`round-trip mismatch: wrote ${JSON.stringify(text)}, read ${JSON.stringify(back)}`);
	}

	// list the dir -> assert the file shows up as a file of the right size.
	const entries = await sandbox.list(".");
	const found = entries.find((f) => f.path === path);
	if (!found) throw new Error(`${path} not in listing: [${entries.map((f) => f.path).join(", ")}]`);
	if (found.isDir) throw new Error(`${path} listed as a directory, expected a file`);
	if (Number(found.size) !== bytes.length) {
		throw new Error(`listed size ${found.size} != ${bytes.length} bytes read`);
	}

	console.log(`✓ files-in-a-sandbox: write → read "${back.trim()}" → list (${entries.length} entry) → dispose`);
} catch (e) {
	console.error("✗ " + (e?.message ?? e));
	process.exitCode = 1;
} finally {
	if (sandbox) {
		try {
			await sandbox[Symbol.asyncDispose]();
		} catch {
			/* self-reaps via idle/lifetime caps */
		}
	}
}

File: examples/smolagents/verify.py
"""
Proves the Tenki-facing half of this example: the `TenkiExecutor` runs
CodeAgent-style Python in a live Tenki sandbox, keeps state across steps, and
detects `final_answer(...)` — the whole `RemotePythonExecutor` seam, asserted
with no LLM. (The full `CodeAgent`, which needs a model key, is in agent.py.)

Needs Python 3.10+ and the deps in requirements.txt. Token/project from env (CI)
or ~/.config/tenki/config.yaml (local `tenki login`). Exits non-zero on failure.
"""
import os
import sys

from smolagents.monitoring import AgentLogger, LogLevel
from smolagents.utils import AgentError

from tenki_executor import TenkiExecutor


def cfg(key):
    try:
        with open(os.path.expanduser("~/.config/tenki/config.yaml")) as f:
            for line in f:
                if line.startswith(key + ":"):
                    return line.split(":", 1)[1].strip()
    except FileNotFoundError:
        pass
    return ""


token = os.environ.get("TENKI_AUTH_TOKEN") or os.environ.get("TENKI_API_KEY") or cfg("auth_token")
if not token:
    print("No token. Set TENKI_AUTH_TOKEN, or run `tenki login`.")
    sys.exit(1)

# Python SDK auth gap: a bare `tenki login` session token must be sent as a
# cookie (prefix `cookie:`); a `tk_` API key works as-is. (Node SDK auto-detects.)
if not token.startswith(("tk_", "ory_st_", "cookie:")):
    token = f"cookie:{token}"

opts = {"auth_token": token, "cpu_cores": 1, "memory_mb": 1024}
if os.environ.get("TENKI_PROJECT_ID") or cfg("current_project_id"):
    opts["project_id"] = os.environ.get("TENKI_PROJECT_ID") or cfg("current_project_id")
if os.environ.get("TENKI_WORKSPACE_ID") or cfg("current_workspace_id"):
    opts["workspace_id"] = os.environ.get("TENKI_WORKSPACE_ID") or cfg("current_workspace_id")

executor = None
try:
    # Build the executor exactly as `CodeAgent(executor=...)` would.
    # LogLevel.OFF keeps CI output to just the ✓/✗ line (step 4 raises on purpose).
    executor = TenkiExecutor(additional_imports=[], logger=AgentLogger(LogLevel.OFF), **opts)

    # 1) Real execution in the microVM.
    out = executor.run_code_raise_errors("print(6 * 7)")
    if out.logs.strip() != "42":
        raise AssertionError(f"exec: expected logs '42', got {out.logs!r}")

    # 2) State persists across separate steps — the reason CodeAgent needs a
    #    live kernel, not a fresh `python3` per call.
    executor.run_code_raise_errors("n = sum(range(11))")
    out = executor.run_code_raise_errors("print(n)")
    if out.logs.strip() != "55":
        raise AssertionError(f"state: expected logs '55', got {out.logs!r}")

    # 3) Final-answer detection. CodeAgent patches its `final_answer` tool to raise
    #    `FinalAnswerException("safe:<json>")`; the executor must catch that and
    #    deserialize the value. We raise that exact wire contract in the sandbox
    #    (no LLM, no tool-package install) and assert the executor surfaces it.
    fa_code = (
        "class FinalAnswerException(BaseException):\n"
        "    def __init__(self, value): self.value = value\n"
        "import json\n"
        "raise FinalAnswerException('safe:' + json.dumps(6 * 7))\n"
    )
    out = executor.run_code_raise_errors(fa_code)
    if not (out.is_final_answer and out.output == 42):
        raise AssertionError(f"final_answer: expected 42, got is_final={out.is_final_answer}, output={out.output!r}")

    # 4) Errors surface as AgentError — what CodeAgent shows the model to retry.
    try:
        executor.run_code_raise_errors("raise ValueError('boom')")
        raise AssertionError("expected AgentError, none raised")
    except AgentError as e:
        if "boom" not in str(e):
            raise AssertionError(f"error path: unexpected message {e!r}")

    print("✓ smolagents: TenkiExecutor ran CodeAgent Python in a Tenki sandbox (42), "
          "kept state across steps (55), caught final_answer (42), surfaced errors")
except Exception as e:  # noqa: BLE001
    print(f"✗ {type(e).__name__}: {e}")
    sys.exit(1)
finally:
    if executor is not None:
        executor.cleanup()
